fixing 6502 analysis

sqaxomonophonen opened the pull request:
36e42b33fe37e922f132e286508bd88ef9fb137d broke analysis by introducing a
new enum value. The analysis code didn't switch on the enum but rather
used it directly as an integer, so since the new `SNES_OP_IMM` has 5 as
its integer value the analysis code would think that the op was 5 bytes
in length.

I have only verified it fixes code analysis on 6502, not SNES. But this patch is fairly simple, and by the looks of it, it was broken on SNES too since it would still assume that `SNES_OP_IMM`'s were 5 bytes in length, not the correct 2 or 3 bytes.

Maijin commented:
Can you add tests in r2r like https://github.com/radare/radare2-regressions/blob/master/t.asm/x86/capstone/x86 ? If you don't you'll probably have to fix stuff again, again and again.

Maijin commented:
For analysis see https://github.com/radare/radare2-regressions/tree/master/t.anal

sqaxomonophonen commented:
My test case was this file: https://veralin.dk/slope.vsf
Try with `r2 slope.vsf -c 's 0x1c64 ; af ; pdf'`
The broken code will "kill" the function analysis after `cmp #0x04` while the fixed code (like the commit before 36e42b33fe37e922f132e286508bd88ef9fb137d) will continue properly until `rts`.

Maijin commented:
if the file is licence free, add it in /bins/ in r2r and add a Pull request with test :)

sqaxomonophonen commented:
It's not license free, but the problem is fairly easy to reproduce so I guess I can make a minimal test case anyway. I don't see a 6502 section in r2r, so should I try to add it to ... /`t.anal/others_anal/`?

sqaxomonophonen commented:
`others_anal` only contains x86 code... should I create a new file or even directory?

Maijin commented:
yep for both

sqaxomonophonen commented:
There's a test now: https://github.com/radare/radare2-regressions/pull/352 (it was wonderfully easy to work with, thanks!)

Maijin commented:
yes is easy, don't forget to check the makefile here https://github.com/radare/radare2-regressions/blob/master/Makefile#L77 and here https://github.com/radare/radare2-regressions/blob/master/Makefile#L67 .

Also regarding tests, looks like your PR break this https://github.com/radare/radare2-regressions/blob/master/t.esil/6502#L2296

See https://s3.amazonaws.com/archive.travis-ci.org/jobs/119602330/log.txt ctrl+f [XX]

sqaxomonophonen commented:
I've updated the Makefile. The failing test was recently changed (https://github.com/radare/radare2-regressions/commit/5a89dec8aff740356d24dda0c49fc056518eaa4f) but it very much looks like a response to the bug introduced in 36e42b33fe37e922f132e286508bd88ef9fb137d and the commit message isn't helpful, so I've reverted the test to its previous state which passes before the bug and after the fix.

radare commented:
Merged!

sqaxomonophonen commented:
Thank you, cheers!
